### PR TITLE
wip: introduce safe callbacks and use in rewriting init manager

### PR DESCRIPTION
**Pull request by mergeconflict**
Introduce a new, "safe" callback mechanism (as in, it's safe to invoke the callback after the receiver has been destroyed), and use it as the basis for a new init manager. See #6116 and #6136 for background.

This is work in progress; I haven't updated use sites yet. Seeking feedback on the basic design.

Signed-off-by: Dan Rosen <mergeconflict@google.com>

### CODE REVIEW

File: source/common/callback/manager.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    return *this;
+  }
+
+  /**
+   * Invokes all callers previously provided to add(). Any corresponding receivers that are no
+   * longer available after invocation (which may have happened as a side-effect of invoking them)
+   * will be removed from the callback manager.
+   * @param args the arguments, if any, to pass to all callers.
+   */
+  void operator()(Args... args) {
```
Comment (htuch) [inline source/common/callback/manager.h]: Do we need to template this? Are args generally super useful to users? Previously we just had `()`.

File: include/envoy/init/init.h [hunk 2]
```diff
@@ -55,9 +38,24 @@ class Manager {
...
   virtual State state() const PURE;
+
+  /**
+   * Register an initialization target. If the manager's current state is uninitialized, the target
+   * will be saved for invocation later, when `initialize` is called. If the current state is
+   * initializing, the target will be invoked immediately. It is an error to register a target with
+   * a manager that is already in initialized state.
+   * @param target_receiver the target to be invoked when initialization begins.
+   */
+  virtual void add(const TargetReceiver& target_receiver) PURE;
```
Comment (htuch) [inline include/envoy/init/init.h]: Is there a constraint that Receiver functors can't have any by-reference capture? Is there a way to enforce that at the type-system level(if they were plain-old C function pointers, that would do that..)?

File: source/common/callback/manager.h (hunk 1 above)
Comment (mergeconflict) [reply source/common/callback/manager.h]: Yes, it's used here. See e.g. `TargetCaller` / `TargetReceiver`. I'm also sort of hoping that this will replace the existing callback mechanism in `source/common/common`, which accepts args.

File: include/envoy/init/init.h (hunk 2 above)
Comment (mergeconflict) [reply include/envoy/init/init.h]: So, yeah, ideally you'd want something like Rust's borrow checker to enforce that your callbacks don't capture something that is destroyed before the callback is invoked. But there doesn't appear to be any such mechanism. So the basic idea behind this design is that the `Receiver` has the same (or narrower) lifetime as any references captured by its callback, which is pretty easy to get right and takes some effort to get wrong.

File: source/common/callback/callback.h [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+// Forward-declaration for Caller's friend declaration.
+template <typename... Args> class Receiver;
+
+/**
+ * Caller: simple wrapper for a weak_ptr to a callback function. Copyable and movable.
+ */
+template <typename... Args> class Caller {
+public:
+  /**
+   * Default constructor for default / value initialization.
```
Comment (htuch) [inline source/common/callback/callback.h]: Why is this needed?

File: source/common/callback/callback.h [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+ * typically should be owned uniquely by the owner of any pointers and references captured by its
+ * handler function. For example, if `this` is captured by the handler function, `this` should
+ * probably also own the Receiver.
+ */
+template <typename... Args> class Receiver {
+public:
+  /**
+   * Default constructor for default / value initialization.
+   */
+  Receiver() = default;
```
Comment (htuch) [inline source/common/callback/callback.h]: Ditto, what happens if this isn't here?

File: source/common/callback/callback.h [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  /**
+   * Reset this receiver, such that any callers previously created will not be able to invoke it.
+   */
+  void reset() { fn_.reset(); }
+
+  /**
+   * Explicit conversion to bool, to test whether the receiver contains a callback function.
+   * @return true if the corresponding Receiver contains a callback, false otherwise.
+   */
+  explicit operator bool() const { return static_cast<bool>(fn_); }
```
Comment (htuch) [inline source/common/callback/callback.h]: Nit: prefer `fn_ != nullptr`.

File: source/common/callback/manager.h (hunk 1 above)
Comment (htuch) [reply source/common/callback/manager.h]: Can you give some examples of where we really need this and can't just get away with passing in a `std::function` that captures its arguments? I think one of the main opportunities to simplify is to reduce the amount of templating here.

File: source/init/callback.cc [hunk 6]
```diff
@@ -0,0 +1,80 @@
+#include "init/callback.h"
```
Comment (htuch) [inline source/init/callback.cc]: This should probably be `source/common/init`, this isn't a top-level Envoy concern.

File: source/init/callback.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * A Caller notifies a Receiver when something has initialized. This works at two levels: first,
+ * each initialization target holds a Caller to notify the init manager's Receiver when the target
+ * has initialized. And second, the manager holds a Caller to notify its client's Receiver when all
+ * of its targets have initialized.
+ *
+ * Caller and Receiver are simple wrappers for their counterparts in Common::Callback with logging
+ * added to help debug initialization and destruction ordering issues that occasionally arise.
+ */
+class Caller : Logger::Loggable<Logger::Id::init> {
```
Comment (htuch) [inline source/init/callback.h]: Wait, why is there a `Caller`/`Receiver` defined here as well as earlier on? :) Can't we just have one abstract version of this?

File: source/common/callback/callback.h (hunk 3 above)
Comment (mergeconflict) [reply source/common/callback/callback.h]: To support the use case where you don't have a Caller until somebody hands one to you, as is the case during initialization:

```c++
class Foo {
public:
  void initialize(Caller<> cb) {
    cb_ = std::move(cb);
    // ...
  }
  void initComplete() {
    cb_();
  }
private:
  Caller<> cb_;
}
```

File: source/common/callback/callback.h (hunk 4 above)
Comment (mergeconflict) [reply source/common/callback/callback.h]: `Receiver` is meant to have similar behavior to both `std::function<...>` and `std::shared_ptr<...>`. You can initialize either of those to be empty, and you can reset them to an empty state after construction, and I figure it's worth preserving that ability here.

File: source/init/callback.cc (hunk 6 above)
Comment (mergeconflict) [reply source/init/callback.cc]: I was wondering about that. The current state of `master` is sort of weird:

* The `Init::Manager` and `Init::Target` interfaces are defined in `include/envoy/init`
* The `InitManagerImpl` class is defined in `source/server` with a TODO comment, "Move InitManagerImpl into a new subdirectory in source/ called init/"

I agree it'd make sense to move the implementation into `source/common/init`. What would you think about also moving the interfaces down into `include/envoy/common/init`?

File: source/init/callback.h (hunk 7 above)
Comment (mergeconflict) [reply source/init/callback.h]: The `Caller`/`Receiver` in common/callback are meant to be as bare-bones as possible, with no logging included. Going on the assumption that we want safer callbacks everywhere, not just in initialization, I didn't want to bake logging as a requirement.

File: source/common/callback/manager.h (hunk 1 above)
Comment (mergeconflict) [reply source/common/callback/manager.h]: I think the example I wrote up in the long comment block in callback.h is a pretty good illustration. Suppose you want to register a bunch of callback functions that consume an event, and pass them the event later.

File: source/init/callback.h (hunk 7 above)
Comment (htuch) [reply source/init/callback.h]: We definitely want to converge these implementations IMHO; even if we have to use macro magic or somethign to do it. They are pretty complicated to reason about, having only a single one is the way to go.

File: source/init/callback.cc (hunk 6 above)
Comment (htuch) [reply source/init/callback.cc]: `source/common/init` for implementation, `include/envoy/init` for interface; that's how Envoy canonical paths work.
